text/0028-generalized-entity-data-source.md

```diff
@@ -0,0 +1,569 @@
+descendants-oriented; `entity_in()` works equally well with both.
+This might be sufficient to support users with descendants-oriented entity data:
+they just need to use the more flexible `EntityDataSource` instead of
+`WholeEntityDataSource`.
+
+### Alternative H
+
+As proposed, `EntityDataSource` methods do not generally allow the
+implementation to assume that input entities exist.
+We could instead allow the implementation to assume that input entities exist,
+which might be more convenient for the implementation.
+However, this would mean that the Cedar evaluator must in some cases call
+`entity_exists()` prior to `entity_attr()` / `entity_in()` / etc, which could
+result in duplicated work in some implementations.
+E.g., many implementations might be able to combine an existence check and
+attribute lookup efficiently, or an existence check and `in` check efficiently.
+
+To complicate this further, for `entity_in()`, the current Cedar evaluator will
+always know that `child` exists, but not necessarily `parent`, prior to calling
+`entity_in()`; but the current RFC proposes not to allow implementations to
+assume this, both for simplicity and to avoid committing to this implementation
+detail.
+
+In the future, deciding to commit to stronger preconditions for
+`EntityDataSource` methods -- e.g., allowing implementations to make stronger
+existence assumptions -- would be a backwards-compatible change, while the
+reverse would not, so in that sense, the current proposal is the safest
+proposal.
+
+### Alternative J
+
+As proposed, `EntityDataSource::entity_in()` requires the implementation to
+consider all the transitive ancestors of the entity, which might be a burden for
+some users.  We could allow users instead to implement
+`entity_parents(&self, uid)`, which returns all the direct parents of the given
+`uid`; and then we could provide a default implementation of `entity_in()` which
+iteratively calls `entity_parents()` if necessary, essentially computing the
+graph reachability in Cedar rather than requiring each `EntityDataSource`
+implementation to compute TC.
+
+## Interactions between this RFC and partial evaluation
+
+Partial evaluation shares some of the motivation of this RFC -- in particular,
+Challenges 1 and 2 in [Motivation](#motivation).
+However, this RFC does not make partial evaluation redundant or vice versa; some
+use-cases will work better by examining Cedar residuals to fetch entity data,
+and other use-cases will prefer to provide an `EntityDataSource` implementation
+for fetching entity data on-demand.
+Some use-cases will even use both in conjunction.
+
+The easiest way to handle the interaction between this RFC and partial
+evaluation would be to leave `EntityDataSource` as currently proposed in this
+RFC, and simply have the Cedar evaluator return a residual as appropriate when
+`EntityDataSource::entity_exists()` returns `false` or
+`WholeEntityDataSource::get()` returns `None`.
```

Comment:
Ack.

Given that the partial evaluation APIs are currently experimental, it is worth thinking about whether we can make the RFC 28 changes more partial-evaluation-aware but still independent enough of partial evaluation that they can be used without the `partial-eval` feature flag.  For instance if `entity_exists()` returns possible Unknown, we'd want the API without `partial-eval` flag to somehow be similar but not have the Unknown option.  This might take some careful design or might not be possible, particularly if we want stabilizing `partial-eval` to be a backwards-compatible change eventually. (AFAIK that is currently the case.)